core/src/main/scala/org/http4s/multipart/Multipart.scala

```diff
@@ -0,0 +1,81 @@
+package org.http4s
+package multipart
+
+import java.io.{ File, FileInputStream, InputStream }
+import java.net.URL
+import java.nio.charset.StandardCharsets
+import java.util.Base64
+import scala.util.Random
+
+import org.http4s._
+import org.http4s.EntityEncoder._
+import org.http4s.MediaType._
+import org.http4s.headers._
+import org.http4s.util.CaseInsensitiveString
+import org.http4s.util.string._
+import scalaz.std.string._
+import scalaz.stream.Process.emit
+import scalaz.stream.text.utf8Encode
+import scodec.bits.{ BitVector, ByteVector }
+
+final case class Part(headers: Headers, body: EntityBody) {
+  def name: Option[CaseInsensitiveString] = headers.get(`Content-Disposition`).map(_.name)
+}
+
+object Part {
+  val empty: Part =
+    Part(Headers.empty, EmptyBody)
+
+  def formData(name: String, value: String, headers: Header*): Part =
+    Part(`Content-Disposition`("form-data", Map("name" -> name)) +: headers, emit(value) |> utf8Encode)
+
+  def fileData(name: String, file: File, headers: Header*): Part = {
+    val in = new FileInputStream(file)
+    try fileData(name, file.getName, in, headers:_*)
+    finally in.close()
+  }
+
+  def fileData(name: String, resource: URL, headers: Header*): Part = {
+    val in = resource.openStream()
+    try fileData(name, resource.getPath.split("/").last, in, headers:_*)
+    finally in.close()
+  }
+
+  private def fileData(name: String, filename: String, in: InputStream, headers: Header*): Part = {
+    // TODO Reading the whole file into one chunk is abominable
+    val body = BitVector.fromInputStream(in).toByteVector
```

Comment:
I tried replacing this with `Process.constant(8192) through scalaz.stream.io.chunkR(in)`, where `in` is passed by name, and MultipartSpec goes into an infinite loop of some sort.  This is probably a bug in MultipartParser, but I don't see it yet.